File: app/nehex-admin/src/services/mail.ts
import { adminFetch } from '@/services/admin-api'

export type MailSmtpSecurity = 'none' | 'starttls' | 'ssl'
export type MailLogStatusFilter = 'all' | 'success' | 'failed'

export type AdminMailSmtpTestPayload = {
  smtp_host: string
  smtp_port: number
  smtp_security: MailSmtpSecurity
  smtp_username?: string | null
  smtp_password?: string | null
  smtp_from_email?: string | null
  smtp_from_name?: string | null
  smtp_timeout_seconds?: number
  test_email: string
}

type AdminActionResponse = {
  message?: unknown
}

export type AdminMailLogItem = {
  id: number
  category: string
  template_key: string
  to_email: string
  subject: string
  body: string
  status: string
  error_message?: string | null
  trigger_comment_id?: number | null
  created_at: string
  sent_at?: string | null
}

type AdminMailLogListResponse = {
  data: AdminMailLogItem[]
  pagination?: {
    page?: number
    size?: number
    total?: number
    total_pages?: number
  }
}

export type AdminMailLogListResult = {
  items: AdminMailLogItem[]
  pagination: {
    page: number
    size: number
    total: number
    total_pages: number
  }
}

async function parseJson<T>(response: Response): Promise<T> {
  return await response.json() as T
}

export async function testAdminMailSmtpConnection(payload: AdminMailSmtpTestPayload): Promise<string> {
  const response = await adminFetch('/admin-api/settings/mail/test', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
  const result = await parseJson<AdminActionResponse>(response)
  return typeof result?.message === 'string' ? result.message : 'SMTP 通信成功'
}

export async function fetchAdminMailLogs(
  status: MailLogStatusFilter,
  page = 1,
  size = 20,
): Promise<AdminMailLogListResult> {
  const params = new URLSearchParams()
  params.set('status', status)
  params.set('page', String(Math.max(1, Math.floor(page))))
  params.set('size', String(Math.max(1, Math.floor(size))))

  const response = await adminFetch(`/admin-api/mail-logs?${params.toString()}`, {
    method: 'GET',
  })
  const payload = await parseJson<AdminMailLogListResponse>(response)
  if (!Array.isArray(payload?.data)) {
    throw new Error('Unexpected mail log response format')
  }

  const safePage = Number.isFinite(payload.pagination?.page) ? Number(payload.pagination?.page) : Math.max(1, page)
  const safeSize = Number.isFinite(payload.pagination?.size) ? Number(payload.pagination?.size) : Math.max(1, size)
  const safeTotal = Number.isFinite(payload.pagination?.total) ? Number(payload.pagination?.total) : payload.data.length
  const safeTotalPages = Number.isFinite(payload.pagination?.total_pages)
    ? Number(payload.pagination?.total_pages)
    : Math.max(0, Math.ceil(safeTotal / safeSize))

  return {
    items: payload.data,
    pagination: {
      page: Math.max(1, safePage),
      size: Math.max(1, safeSize),
      total: Math.max(0, safeTotal),
      total_pages: Math.max(0, safeTotalPages),
    },
  }
}

File: app/services/admin_service_parts/mail.py
from __future__ import annotations

from typing import Any

from sqlalchemy.orm import Session

from app.models.mail_log import MailLog
from app.schemas.admin import AdminMailLogItem
from app.services.mail_service import list_mail_logs, send_mail_test_with_config


def _map_admin_mail_log_item(row: MailLog) -> AdminMailLogItem:
    return AdminMailLogItem(
        id=row.id,
        category=row.category,
        template_key=row.template_key,
        to_email=row.to_email,
        subject=row.subject,
        body=row.body,
        status=row.status.value,
        error_message=row.error_message,
        trigger_comment_id=row.trigger_comment_id,
        created_at=row.created_at,
        sent_at=row.sent_at,
    )


def list_admin_mail_logs(
    session: Session,
    *,
    status: str = "all",
    page: int = 1,
    size: int = 20,
) -> tuple[list[AdminMailLogItem], int, int, int, int]:
    rows, normalized_page, normalized_size, total, total_pages = list_mail_logs(
        session=session,
        status=status,
        page=page,
        size=size,
    )
    return (
        [_map_admin_mail_log_item(row) for row in rows],
        normalized_page,
        normalized_size,
        total,
        total_pages,
    )


def test_admin_mail_smtp(session: Session, payload: dict[str, Any]) -> None:
    send_mail_test_with_config(
        session=session,
        smtp_host=payload.get("smtp_host"),
        smtp_port=payload.get("smtp_port"),
        smtp_security=payload.get("smtp_security"),
        smtp_username=payload.get("smtp_username"),
        smtp_password=payload.get("smtp_password"),
        smtp_from_email=payload.get("smtp_from_email"),
        smtp_from_name=payload.get("smtp_from_name"),
        smtp_timeout_seconds=payload.get("smtp_timeout_seconds"),
        test_email=payload.get("test_email") or "",
    )
